Dockerfile

```diff
@@ -0,0 +1,32 @@
+FROM python:3.11-slim-bullseye
+
+ENV DEBIAN_FRONTEND noninteractive
+ENV TERM linux
+
+# Install distribution packages, with caching.
+# `git` is needed for `versioningit`.
+RUN rm -f /etc/apt/apt.conf.d/docker-clean; echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' > /etc/apt/apt.conf.d/keep-cache
+RUN \
+    --mount=type=cache,id=apt,sharing=locked,target=/var/cache/apt \
+    --mount=type=cache,id=apt,sharing=locked,target=/var/lib/apt \
+    true \
+    && apt-get update \
+    && apt-get install --no-install-recommends --no-install-suggests --yes git firefox-esr xvfb
+
+# Copy sources
+COPY . /src
+
+# Install package, with caching of dependency packages.
+RUN \
+    --mount=type=cache,id=pip,target=/root/.cache/pip \
+    true \
+    && pip install --use-pep517 --prefer-binary '/src'
+
+# Uninstall Git again.
+RUN apt-get --yes remove --purge git && apt-get --yes autoremove
+
+# Purge /src and /tmp directories.
+RUN rm -rf /src /tmp/*
+
+ENV PATH /usr/lib/firefox-esr:/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin
+CMD xvfb-run python -c 'from grafanimate.commands import run; run()'
```

Comment:
@intermittentnrg: I am struggling here to find an incantation to let `xvfb-run grafanimate` be the baseline command, and "just add parameters" when invoking the Docker container.

```
#CMD [ "xvfb-run", "grafanimate" ]
#ENTRYPOINT [ "xvfb-run", "python -c 'from grafanimate.commands import run; run()'" ]
#ENTRYPOINT [ "xvfb-run" ]
```

I haven't been successful so far. It always blocks.

```
docker build --tag=local/grafanimate --file=Dockerfile .
docker run --rm -it local/grafanimate
```

Reply:
On the other hand, I actually don't like obfuscating the Docker entrypoint command to level that will prevent being able to just run `bash`, in order to explore the ingredients.

Reply:
I tried using ```ENTRYPOINT ["xvfb-run", "python", "-c", "from grafanimate.commands import run; run()"]``` but it just froze without running python command.

Changing it to ```ENTRYPOINT xvfb-run python -c "from grafanimate.commands import run; run()" "$@"``` works .. not sure why I think the former is better. It should be the same just prefixed by `sh -c`

Reply:
It works well when using:
```
ENTRYPOINT [ "grafanimate" ]
```

It blocks when using:
```
ENTRYPOINT [ "xvfb-run", "grafanimate" ]
```

Reply:
See my comment https://github.com/panodata/grafanimate/pull/21#discussion_r1361341726 lol

Reply:
I don't see how that will work. Try running `docker run --rm -it local/grafanimate --version`.

Reply:
```
$ docker run --rm -it grafanimate --version
Error: Parameter --scenario is mandatory
Usage:
      grafanimate [options] [--target=<target>]...
      grafanimate --version
      grafanimate (-h | --help)
```
I git stash'd so shouldn't be anything uncommited. WorksForMe?

Reply:
```
$ docker run --rm -it grafanimate --version
Error: Parameter --scenario is mandatory
Usage:
      grafanimate [options] [--target=<target>]...
      grafanimate --version
      grafanimate (-h | --help)
```
I git stash'd and docker build so shouldn't be anything uncommited. WorksForMe?

Reply:
I see that it works as you described at https://github.com/panodata/grafanimate/pull/21#discussion_r1361348752. I will need to investigate why just invoking `--version` does not work on my end.

When supplying the `--scenario` option grafanimate is asking for, it apparently also seems to work here.

But why does the most basic incantation, `grafanimate --version` not work?

Reply:
The output of `grafanimate --version` should look like this:
```
$ grafanimate --version
grafanimate 0.7.0
```

Reply:
See https://github.com/panodata/grafanimate/pull/21#discussion_r1361352470

Not sure why --version isn't handled right. Maybe docker/xvffb-run/python also handles `--version` and takes it but no it doesn't make sense when other options work...

This does work without `--`, by putting --version after other options.

```
$ docker run --rm grafanimate --header-layout=no-chrome --scenario=scenarios.py:gb_windsolar --output=./animations --video-fps=30 --video-framerate=30 --exposure-time=0.2 --version
grafanimate 0.7.0
```

Reply:
It works perfectly when using
```
ENTRYPOINT [ "grafanimate" ]
```

```
$ docker run --rm local/grafanimate --version
grafanimate 0.7.0
```

Reply:
When interactively logging into the container, running `xvfb-run grafanimate --version` also works well.

Reply:
Just needs `xvfb-run` wrapper for headless X11 server to run firefox.

If you get grafanimate command to work then that's preferrable over `python -c 'from grafanimate.commands import run; run()'`.

As I mentioned I don't code in python normally..

Reply:
Just needs `xvfb-run` wrapper for headless X11 server to run firefox.

If you get grafanimate command to work then that's preferrable over `python -c 'from grafanimate.commands import run; run()'`.

As I mentioned I don't code in python normally..


Yes for my first PR i only tested it interactively and pasted command that worked as ENTRYPOINT and then we started having issues...

Reply:
Says to run `docker run --init` here https://stackoverflow.com/a/72017110/34516

```
      --init                           Run an init inside the container that forwards signals and reaps processes
```

Also X11/firefox may like having a home directory in the container, and not run as root. But these things are not blockers.

Reply:
You can also try this command to recognize something is not quite right:
```
$ docker run --rm local/grafanimate --scenario=scenarios.py:gb_windsolar
Error: Parameter --scenario is mandatory
```

Reply:
> looks like running the command withouth init parameter, running as PID 1, is not handling the signal USR1 correctly.
https://stackoverflow.com/a/71753843/34516

Maybe xvfb-run doesn't like to run as pid 1. And therefore not using the [] syntax for entrypoint spawns shell process that runs as pid 1.

Reply:
I'm in Sweden, same timezone. Thanks! let's continue later.

Reply:
I'm in Sweden, same timezone. Thanks! let's continue later.

Reply:
I am not insisting on the `[]` syntax, I am also currently using your proposal.
```
ENTRYPOINT xvfb-run python -c "from grafanimate.commands import run; run()" "$@"
```
But unfortunately, it does not work yet. The first command-line option is not propagated properly.

Reply:
This works:
```
ENTRYPOINT ["xvfb-run", "python", "-c", "from grafanimate.commands import run; run()"]
```
```
$ docker run --rm -ti --init grafanimate --version
grafanimate 0.7.0
```

In my PR #21 I added tini per instructions in 03be161. docker run --init also uses tini.

In f06bae8 I also made setup.py cached by docker build cache so it doesn't re-run every time you change a file and run docker build.

Reply:
Wonderful. At [^1], I discovered a recipe to inline that, so that the user does not need to specify `--init` as a command-line option to `docker run`.

```Dockerfile
# Add Tini.
ENV TINI_VERSION v0.19.0
ADD https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini /tini
RUN chmod +x /tini

# Run grafananimate through tini and xvfb-run.
ENTRYPOINT ["/tini", "--", "xvfb-run", "grafanimate"]
```

[^1]: https://github.com/krallin/tini#using-tini

Reply:
Yes this is what I did in 03be161